pkg/serverless/trigger/events/events.go

```diff
@@ -343,17 +343,74 @@ type LambdaFunctionURLRequestContextHTTPDescription struct {
 // StepFunctionEvent is the event you get when you instrument a legacy Stepfunction Lambda:Invoke task state
 // as recommended by https://docs.datadoghq.com/serverless/step_functions/installation?tab=custom
 // This isn't an "official" event, as a default StepFunction invocation will just contain {}
-type StepFunctionEvent struct {
-	Payload StepFunctionPayload
+type StepFunctionEvent[T any] struct {
+	Payload T
 }
 
 // StepFunctionPayload is the payload of a StepFunctionEvent. It's also a non-legacy version of the `StepFunctionEvent`.
 type StepFunctionPayload struct {
 	Execution struct {
-		ID string
+		ID           string
+		RedriveCount uint16
 	}
 	State struct {
 		Name        string
 		EnteredTime string
+		RetryCount  uint16
 	}
 }
+
+// NestedStepFunctionPayload contains a StepFunctionPayload but also has the Execution ID of the top-most Step Function in the trace
+type NestedStepFunctionPayload struct {
+	Payload           StepFunctionPayload
+	RootExecutionID   string
+	ServerlessVersion string
+}
+
+// LambdaRootStepFunctionPayload contains a StepFunctionPayload but also has the Trace ID of the top-most Lambda in the trace
+type LambdaRootStepFunctionPayload struct {
+	Payload           StepFunctionPayload
+	TraceID           string
+	TraceTags         string
+	ServerlessVersion string
+}
+
+// genericUnmarshal helps extract fields from _datadog.
+func genericUnmarshal(data []byte, fieldMap map[string]interface{}) error {
```

Comment:
The custom unmarshaler is a consequence of how I defined the types above.

For example, the `StepFunctionPayload` should be at the same top-level as the `RootExecutionID` and `ServerlessVersion` for `type NestedStepFunctionPayload`. But I liked this nested approach so that we can deal with this shared context object the exact same way in all cases in `carriers.go`.

We can simply pass the whole payload into `extractTraceContextFromStepFunctionContext()` without needing to deal with each case seperately